data_mgr.py

```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env python3
+
+# -*- coding: utf-8 -*-
+# Copyright (C) 2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Create and update the data structure for all workflow services."""
+
+from fnmatch import fnmatchcase
+
+from tornado import gen
+# from google.protobuf.json_format import MessageToDict
+
+# from cylc import flags
+from cylc.exceptions import ClientError, ClientTimeout
+from cylc.ws_messages_pb2 import PbEntireWorkflow
+from cylc.network.scan import MSG_TIMEOUT
+
+
+METHOD_MAP = {
+    'pb_entire_workflow': PbEntireWorkflow
+}
+
+
+async def get_workflow_data(w_id, client, method):
+    # Use already established client
+    try:
+        pb_msg = await client.async_request(method)
+    except ClientTimeout as exc:
+        return (w_id, MSG_TIMEOUT)
+    except ClientError as exc:
+        return (w_id, None)
+    else:
+        ws_data = METHOD_MAP[method]()
```

Comment:
We may/will end up using more than just one method in the request:
```
async def get_workflow_data(w_id, client, method):
    # Use already established client
    try:
        pb_msg = await client.async_request(method)
    except ClientTimeout as exc:
        return (w_id, MSG_TIMEOUT)
    except ClientError as exc:
        return (w_id, None)
    else:
        ws_data = METHOD_MAP[method]()
        ws_data.ParseFromString(pb_msg)
    return (w_id, ws_data)
```

And I need to map the method to the protobuf message... I need this because I cannot read the bytes content without know what it is.

Reply:
Yup, but do we need the protobuf message type as text, instead of the Python object? i.e., instead of `get_workflow_data` receiving a `client` (which I believe is a `: cylc.flow.network.client.SuiteRuntimeClient` instance), and a `method: string`, this last one, could it be `method: object`? So instead of passing `"PbPoolTask"`, I would have to `from ws_messages_pb2 import PbPoolTask` and pass the object `PbPoolTask`. If so, we wouldn't need to dict with `text -> object` mapping?